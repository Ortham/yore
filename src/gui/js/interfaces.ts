export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface Photo {
  path: string;
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
  height?: number;
  width?: number;
  loaded?: boolean;
}

interface Location {
  Existing?: Coordinates;
  Suggested?: [Coordinates, LocationAccuracy];
}

export interface LocationAccuracy {
  meters: number;
  seconds: number;
}

export interface LocationResponse {
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
}
