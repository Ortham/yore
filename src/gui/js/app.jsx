import 'babel-polyfill';
import 'whatwg-fetch';
import React from 'react';
import ReactDOM from 'react-dom';
import Page from './page';
import { getPhotos, getRootPath } from './requests';

function getInitialState() {
  const state = {
    rootPath: undefined,
    filterPhotos: false,
    currentPhoto: undefined,
    photos: []
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath;

      return getPhotos();
    })
    .then(photos => {
      state.photos = photos;

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(
    <Page rootPath={state.rootPath} photos={state.photos} />,
    document.getElementById('root')
  );
});
