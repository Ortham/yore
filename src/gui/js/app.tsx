import '@babel/polyfill';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import 'whatwg-fetch';
import { Page, PageProps } from './page';
import {
  getInterpolate,
  getLocationHistoryPath,
  getPhotos,
  getRootPath,
  getLocations
} from './requests';
import { updatePhotoLocations } from './photo';

function getInitialState() {
  const state: PageProps = {
    interpolate: false,
    locationHistoryPath: undefined,
    photos: [],
    rootPath: undefined
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath || '';

      return getPhotos();
    })
    .then(photos =>
      getLocations(0, photos.length).then(locations =>
        updatePhotoLocations(photos, locations)
      )
    )
    .then(photos => {
      state.photos = photos;

      return getInterpolate();
    })
    .then(responseBody => {
      state.interpolate = responseBody.interpolate;

      return getLocationHistoryPath();
    })
    .then(responseBody => {
      state.locationHistoryPath = responseBody.locationHistoryPath || '';

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(<Page {...state} />, document.getElementById('root'));
});
