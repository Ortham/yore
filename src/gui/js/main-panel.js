import React from 'react';
import MapArea from './map-area';
import {locationDescription, hasSuggestion} from './photo';

export default function MainPanel(props) {
    if (props.photo) {
        return (
            <main>
                <section>
                    <img src={props.photo.src} />
                    <MapArea photo={props.photo} />
                </section>
                <footer>
                    {locationDescription(props.photo)}
                    <br />
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionApply} >
                        Apply
                    </button>
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionDiscard} >
                        Discard
                    </button>
                </footer>
            </main>
        );
    } else {
        return (
            <main>
                <section>No photo selected.</section>
                <footer>
                    <br />
                    <button disabled>Apply</button>
                    <button disabled>Discard</button>
                </footer>
            </main>
        );
    }
}
