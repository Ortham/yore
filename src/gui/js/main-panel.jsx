import React from 'react';
import PropTypes from 'prop-types';
import MapArea from './map-area';
import { locationDescription, hasSuggestion } from './photo';

export default function MainPanel(props) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}

MainPanel.propTypes = {
  photo: PropTypes.shape({
    src: PropTypes.string
  }),
  handleSuggestionApply: PropTypes.func.isRequired,
  handleSuggestionDiscard: PropTypes.func.isRequired
};

MainPanel.defaultProps = {
  photo: undefined
};
