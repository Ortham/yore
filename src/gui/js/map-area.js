import React from 'react';
import GoogleMapReact from 'google-map-react';
import FaMapMarker from 'react-icons/lib/fa/map-marker';
import {googleMapsCoordinates} from './photo';

function MapMarker(props) {
    return <FaMapMarker lat={props.lat} lng={props.lng} style={{height: '24px', width: '24px', color: 'crimson'}} />;
}

export default function MapArea(props) {
    if (props.photo && props.photo.location) {
        const coord = googleMapsCoordinates(props.photo);
        return (
            <GoogleMapReact center={coord} zoom={5}>
                <MapMarker lat={coord.lat} lng={coord.lng} />
            </GoogleMapReact>
        );
    } else {
        return <GoogleMapReact center={[0, 0]} zoom={5} />;
    }
}
