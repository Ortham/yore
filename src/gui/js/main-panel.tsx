import * as React from 'react';
import { Photo } from './interfaces';
import { MapArea } from './map-area';
import { hasSuggestion, locationDescription } from './photo';

export interface MainPanelProps {
  photo?: Photo;
  handleSuggestionApply: () => Promise<void>;
  handleSuggestionDiscard: () => void;
}

export function MainPanel(props: MainPanelProps) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}
