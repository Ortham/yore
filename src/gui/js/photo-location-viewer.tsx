import * as React from 'react';
import { Photo } from './interfaces';
import { MapArea } from './map-area';
import { hasSuggestion, locationDescription } from './photo';

function getPhotoSrc(photo: Photo) {
  return `/thumbnail?path=${encodeURIComponent(
    photo.path
  )}&maxWidth=300&maxHeight=300`;
}

export interface PhotoLocationViewerProps {
  photo?: Photo;
  handleSuggestionApply: () => Promise<void>;
  handleSuggestionDiscard: () => void;
}

export function PhotoLocationViewer(props: PhotoLocationViewerProps) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={getPhotoSrc(props.photo)} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              type="button"
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              type="button"
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button type="button" disabled>
            Apply
          </button>
          <button type="button" disabled>
            Discard
          </button>
        </div>
      </footer>
    </div>
  );
}
