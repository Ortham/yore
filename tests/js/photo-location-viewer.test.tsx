import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Photo, LocationAccuracy } from '../../src/gui/js/interfaces';

jest.mock('../../src/gui/js/map-area', () => ({
  MapArea: 'MapArea'
}));

import { PhotoLocationViewer } from '../../src/gui/js/photo-location-viewer'; // eslint-disable-line import/first

function apply() {
  return Promise.resolve();
}

function discard() {}

describe('PhotoLocationViewer', () => {
  test('renders an image, map, location description and disabled buttons if photo location is not suggested', () => {
    const photo: Photo = {
      path: 'path',
      location: {
        Existing: {
          latitude: 52.0,
          longitude: 36.2
        }
      }
    };
    const photoLocationViewer = renderer
      .create(
        <PhotoLocationViewer
          photo={photo}
          handleSuggestionApply={apply}
          handleSuggestionDiscard={discard}
        />
      )
      .toJSON();
    expect(photoLocationViewer).toMatchSnapshot();
  });

  test('renders an image, map, text and disabled buttons if photo has no location', () => {
    const photo = {
      path: 'path'
    };
    const photoLocationViewer = renderer
      .create(
        <PhotoLocationViewer
          photo={photo}
          handleSuggestionApply={apply}
          handleSuggestionDiscard={discard}
        />
      )
      .toJSON();
    expect(photoLocationViewer).toMatchSnapshot();
  });

  test('renders an image, map, location description and enabled buttons if photo location is suggested', () => {
    const photo = {
      path: 'path',
      location: {
        Suggested: [
          {
            latitude: 52.0,
            longitude: 36.2
          },
          {
            meters: 5,
            seconds: 10
          }
        ] as [Coordinates, LocationAccuracy]
      }
    };
    const photoLocationViewer = renderer
      .create(
        <PhotoLocationViewer
          photo={photo}
          handleSuggestionApply={apply}
          handleSuggestionDiscard={discard}
        />
      )
      .toJSON();
    expect(photoLocationViewer).toMatchSnapshot();
  });

  test('renders text and disabled buttons if photo is undefined', () => {
    const photoLocationViewer = renderer
      .create(
        <PhotoLocationViewer
          handleSuggestionApply={apply}
          handleSuggestionDiscard={discard}
        />
      )
      .toJSON();
    expect(photoLocationViewer).toMatchSnapshot();
  });
});
