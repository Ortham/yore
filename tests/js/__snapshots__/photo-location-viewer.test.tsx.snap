// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoLocationViewer renders an image, map, location description and disabled buttons if photo location is not suggested 1`] = `
<div
  id="main"
>
  <section>
    <img
      alt="Selected"
      src="path"
    />
    <MapArea
      photo={
        Object {
          "location": Object {
            "Existing": Object {
              "latitude": 52,
              "longitude": 36.2,
            },
          },
          "path": "",
          "src": "path",
        }
      }
    />
  </section>
  <footer>
    <div>
      Existing location
    </div>
    <div>
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Apply
      </button>
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Discard
      </button>
    </div>
  </footer>
</div>
`;

exports[`PhotoLocationViewer renders an image, map, location description and enabled buttons if photo location is suggested 1`] = `
<div
  id="main"
>
  <section>
    <img
      alt="Selected"
      src="path"
    />
    <MapArea
      photo={
        Object {
          "location": Object {
            "Suggested": Array [
              Object {
                "latitude": 52,
                "longitude": 36.2,
              },
              Object {
                "meters": 5,
                "seconds": 10,
              },
            ],
          },
          "path": "",
          "src": "path",
        }
      }
    />
  </section>
  <footer>
    <div>
      Suggested location: accuracy is 5 meters and 10 seconds
    </div>
    <div>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Apply
      </button>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Discard
      </button>
    </div>
  </footer>
</div>
`;

exports[`PhotoLocationViewer renders an image, map, text and disabled buttons if photo has no location 1`] = `
<div
  id="main"
>
  <section>
    <img
      alt="Selected"
      src="path"
    />
    <MapArea
      photo={
        Object {
          "path": "",
          "src": "path",
        }
      }
    />
  </section>
  <footer>
    <div>
      No location
    </div>
    <div>
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Apply
      </button>
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Discard
      </button>
    </div>
  </footer>
</div>
`;

exports[`PhotoLocationViewer renders text and disabled buttons if photo is undefined 1`] = `
<div
  id="main"
>
  <section>
    No photo selected.
  </section>
  <footer>
    <div />
    <div>
      <button
        disabled={true}
        type="button"
      >
        Apply
      </button>
      <button
        disabled={true}
        type="button"
      >
        Discard
      </button>
    </div>
  </footer>
</div>
`;
